### PR TITLE
remove device_id parameter out of ExecutionProvider::GetAllocator()

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
Remove the parameter device_id out of ExecutionProvider::GetAllocator() function

### Motivation and Context
The parameter device_id is not necessary. We can fully rely on the second parameter OrtMemType mem_type to determine the device_id when getting allocator from executionProvider.

### CODE REVIEW

**Comment by souptc** [comment]
LGTM, but let's confirm all the EP implement GetDevice correctly.

File: include/onnxruntime/core/framework/execution_provider.h [hunk 1]
```diff
@@ -77,7 +77,7 @@ class IExecutionProvider {
   /**
    * Get an allocator with specified device id and MemType. Return nullptr if it doesn't exist
    */
-  virtual AllocatorPtr GetAllocator(int device_id, OrtMemType mem_type) const;
+  virtual AllocatorPtr GetAllocator(OrtMemType mem_type) const;
```
Comment (pranavsharma) [inline include/onnxruntime/core/framework/execution_provider.h:80]: The removal of device_id makes this method a bit misleading. It gives me the impression that the output of GetAllocator depends on OrtMemType only. As it stands now it's very clear that an allocator is a function of both device_id and mem_type. What's the disadvantage of keeping the status quo?

Comment (pranavsharma) [reply include/onnxruntime/core/framework/execution_provider.h:80]: Alright, given that an EP is associated with a device I suppose this is ok. This does mean that GetDevice() *must* be implemented.

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**